File: csharp/CoursioApi.cs
using System;
using System.Net;
using System.Text;
using System.IO;
using System.Security.Cryptography;

namespace CoursioApi
{
	public class CoursioApi
	{
		protected WebRequest request;

		protected string salt;
		protected string publicKey;
		protected string privateKey;
		protected string baseUrl = "https://api.coursio.com/v1/";

		public CoursioApi (string publicKey, string privateKey, string salt = "coursiosalt")
		{
			if (publicKey == null || privateKey == null)
			{
				throw new System.Exception ("Both public and private key are required.");
			}

			this.salt = salt;
			this.publicKey = publicKey;
			this.privateKey = privateKey;
		}

		public string Get(string endpoint, int objectId = 0)
		{
			if (objectId != 0)
			{
				endpoint += "/" + objectId.ToString();
			}

			Prepare (endpoint);
			request.Method = "GET";

			return ReadResponse ();
		}

		public string Post(string endpoint, string jsonString)
		{
			Prepare (endpoint);
			request.Method = "POST";

			byte[] byteArray = Encoding.UTF8.GetBytes (jsonString);

			// Set the ContentLength property of the WebRequest.
			request.ContentLength = byteArray.Length;

			// Write data to the Stream
			Stream dataStream = request.GetRequestStream ();
			dataStream.Write (byteArray, 0, byteArray.Length);
			dataStream.Close ();

			return ReadResponse ();
		}

		public string Put(string endpoint, int objectId, string jsonString)
		{
			if (objectId == 0)
			{
				throw new System.Exception ("No object ID specified");
			}
			endpoint += "/" + objectId.ToString();

			Prepare (endpoint);
			request.Method = "PUT";

			byte[] byteArray = Encoding.UTF8.GetBytes (jsonString);

			// Set the ContentLength property of the WebRequest.
			request.ContentLength = byteArray.Length;

			// Write data to the Stream
			Stream dataStream = request.GetRequestStream ();
			dataStream.Write (byteArray, 0, byteArray.Length);
			dataStream.Close ();

			return ReadResponse ();
		}

		public string Delete(string endpoint, int objectId)
		{
			if (objectId == 0)
			{
				throw new System.Exception ("No object ID specified");
			}
			endpoint += "/" + objectId.ToString();

			Prepare (endpoint);
			request.Method = "DELETE";

			return ReadResponse ();
		}

		protected void Prepare(string endpoint)
		{
			if (endpoint == null || endpoint.Length == 0)
			{
				throw new System.Exception ("No endpoint specified");
			}

			// setup connection to endpoint
			request = WebRequest.Create(baseUrl + endpoint);

			// compute HMAC
			var enc = Encoding.ASCII;
			HMACSHA1 hmac = new HMACSHA1(enc.GetBytes(privateKey));
			hmac.Initialize();

			var timestamp = DateTime.Now.ToString(@"MM\/dd\/yyyy h\:mm tt"); 
			var signatureString = publicKey + timestamp + salt;
			byte[] buffer = enc.GetBytes(signatureString);
			var hash = BitConverter.ToString(hmac.ComputeHash(buffer)).Replace("-", "").ToLower();

			request.Headers ["X-Coursio-apikey"] = publicKey;
			request.Headers ["X-Coursio-time"] = timestamp;
			request.Headers ["X-Coursio-random"] = salt;
			request.Headers ["X-Coursio-hmac"] = hash;
		}

		protected string ReadResponse()
		{
			// Get the response.
			WebResponse response = request.GetResponse ();

			//	Console.WriteLine (((HttpWebResponse)response).StatusDescription);

			// Get the stream content and read it
			Stream dataStream = response.GetResponseStream ();
			StreamReader reader = new StreamReader (dataStream);

			// Read the content.
			string responseFromServer = reader.ReadToEnd ();

			// Clean up
			reader.Close ();
			dataStream.Close ();
			response.Close ();

			return responseFromServer;
		}
	}
}

File: php/CoursioApi.php
<?php

class CoursioApi
{
    protected $request;

    protected $salt;
    protected $publicKey;
    protected $privateKey;
    protected $baseUrl = 'https://api.coursio.com/v1/';

    public function __construct($publicKey, $privateKey, $salt = 'coursio_salt')
    {
        if (!$publicKey || !$privateKey)
        {
            throw new Exception('Both public and private key are required.');
        }

        $this->salt = $salt;
        $this->publicKey = $publicKey;
        $this->privateKey = $privateKey;
    }

    /**
     * API GET wrapper
     *
     * @param $endpoint
     * @param int $objectId
     * @return mixed
     */
    public function Get($endpoint, $objectId = 0)
    {
        if ($objectId)
        {
            $endpoint .= '/' . $objectId;
        }
        $this->Prepare($endpoint);

        curl_setopt($this->request, CURLOPT_CUSTOMREQUEST, 'GET');

        return $this->ReadResponse();
    }

    /**
     * API POST wrapper
     *
     * @param $endpoint
     * @param $data
     * @return mixed
     */
    public function Post($endpoint, $data)
    {
        $this->Prepare($endpoint);

        curl_setopt($this->request, CURLOPT_CUSTOMREQUEST, 'POST');
        curl_setopt($this->request, CURLOPT_POSTFIELDS, json_encode($data));

        return $this->ReadResponse();
    }

    /**
     * API PUT wrapper
     *
     * @param $endpoint
     * @param $objectId
     * @param $data
     * @return mixed
     */
    public function Put($endpoint, $objectId, $data)
    {
        if ($objectId)
        {
            $endpoint .= '/' . $objectId;
        }
        $this->Prepare($endpoint);

        curl_setopt($this->request, CURLOPT_CUSTOMREQUEST, 'PUT');
        curl_setopt($this->request, CURLOPT_POSTFIELDS, json_encode($data));

        return $this->ReadResponse();
    }

    /**
     * API DELETE wrapper
     *
     * @param $endpoint
     * @param int $objectId
     * @return mixed
     * @throws Exception
     */
    public function Delete($endpoint, $objectId = 0)
    {
        if (!$objectId)
        {
            throw new Exception('Object ID is required');
        }

        $this->Prepare($endpoint . '/' . $objectId);

        curl_setopt($this->request, CURLOPT_CUSTOMREQUEST, 'DELETE');

        return $this->ReadResponse();
    }

    /**
     * Prepares the API call, sets the headers
     *
     * @param $endpoint
     * @throws Exception
     */
    protected function Prepare($endpoint)
    {
        if (!$endpoint || !strlen($endpoint))
        {
            throw new Exception('No endpoint specified');
        }

        // compute HMAC
        $timestamp = microtime(true);
        $rawString = $this->publicKey . $timestamp . $this->salt;
        $hash = hash_hmac ('sha1', $rawString , $this->privateKey);

        $headers = array
        (
            'X-Coursio-apikey: ' . $this->publicKey,
            'X-Coursio-time: ' . $timestamp,
            'X-Coursio-random: ' . $this->salt,
            'X-Coursio-hmac: ' . $hash,
        );

        // setup connection to endpoint
        $this->request = curl_init($this->baseUrl . $endpoint);
        curl_setopt($this->request, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($this->request, CURLOPT_RETURNTRANSFER, true);
    }

    /**
     * Reads the API response and decodes it
     *
     * @return mixed
     */
    protected function ReadResponse()
    {
        $result = curl_exec($this->request);
        curl_close($this->request);

        return json_decode($result, true);
	}
}
